---
title: "16s_remove primers"
output:
  word_document: default
 
---

# -----------------------------------------------------------------------------#
# V6V8
# remove adaptors
# Author: Geoffrey Zahn
# Software versions:  R v 4.2.2
#                     tidyverse v 1.3.2
#                     dada2 v 1.24.0
#                     decontam v 1.6.0
#                     Biostrings v 2.64.0
#                     ShortRead v 1.54.0
# -----------------------------------------------------------------------------#

#############################################################
#### This script calls cutadapt to remove any primers    ####
#### You must have cutadapt installed on your system     ####
#### and present in your PATH. See cutadapt installation ####
#### documents for instructions.                         ####
#############################################################

# PACKAGES, SCRIPTS, AND SETUP ####

setwd("C:/Users/becca/Dropbox/data working/16S_soils/data")
```{r}

library(tidyverse); packageVersion("tidyverse")
library(dada2); packageVersion("dada2")
library(Biostrings); packageVersion("Biostrings")
library(ShortRead); packageVersion("ShortRead")
```


# PARSE FILE PATHS ####

# File parsing

```{r}
path <- "C:/Users/becca/data" # CHANGE to the subdirectory containing your demultiplexed fastq files

fnFs <- sort(list.files(path, pattern = "_L001_R1_001", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_L001_R2_001" , full.names = TRUE))
```


# CHECK FOR AND REMOVE PRIMER SITES WITH CUTADAPT ####
```{r}
# Here, you should supply the primer sequences used during PCR

# to amplify the V6$8 region of bacterial 16S rDNA
FWD <- "ACGCGHNRAACCTTACC" # Sequence of FWD primer
REV <- "ACGGGCRGTGWGTRCAA"  # Sequence of REV primer
# you can find the absolute file path to your cutadapt program in your terminal by entering "which cutadapt"
# and then copying the output here
```


which cutadapt

Set up 
######################################################################################################

```{r}
allOrients <- function(primer) {
  # Create all orientations of the input sequence
  require(Biostrings)
  dna <- DNAString(primer)  # The Biostrings works w/ DNAString objects rather than character vectors
  orients <- c(Forward = dna, Complement = complement(dna), Reverse = reverse(dna), 
               RevComp = reverseComplement(dna))
  return(sapply(orients, toString))  # Convert back to character vector
}


FWD.orients <- allOrients(FWD)
REV.orients <- allOrients(REV)
FWD.orients; REV.orients
```

Prefilter to remove reads with ambiguous (N) bases ####
```{r}
fnFs.filtN <- file.path(path,"filtN",basename(fnFs)) # Put N-filterd files in filtN/ subdirectory
fnRs.filtN <- file.path(path,"filtN", basename(fnRs))
filterAndTrim(fnFs, fnFs.filtN, fnRs, fnRs.filtN, maxN = 0, multithread = F)
```


# Discover primer matches, regardless of orientation ####
```{r}
primerHits <- function(primer, fn) {
  # Counts number of reads in which the primer is found
  nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE)
  return(sum(nhits > 0))
}

rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs.filtN[[1]]), 
      FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = fnRs.filtN[[1]]), 
      REV.ForwardReads = sapply(REV.orients, primerHits, fn = fnFs.filtN[[1]]), 
      REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnRs.filtN[[1]]))

```


# Run cutadapt ####

```{r}
cutadapt <- path.expand("C:/Users/becca/cutadapt")
system2(cutadapt, args = "--version") # If returns an error, you do not have cutadapt installed correctly
```


```{r}
path.cut <- file.path(path, cutadapt)
if(!dir.exists(path.cut)) dir.create(path.cut)
fnFs.cut <- file.path(path.cut, basename(fnFs))
fnRs.cut <- file.path(path.cut, basename(fnRs))

FWD.RC <- dada2:::rc(FWD)
REV.RC <- dada2:::rc(REV)

```

```{r}
# Trim FWD and the reverse-complement of REV off of R1 (forward reads)
R1.flags <- paste("-g", FWD, "-a", REV.RC) 
# Trim REV and the reverse-complement of FWD off of R2 (reverse reads)
R2.flags <- paste("-G", REV, "-A", FWD.RC)

```

-g: sequence to trim off the 5’ end of the forward read (forward primer) -a: sequence to trim off the 3’ end of the forward read (reverse complemented reverse primer) -G: sequence to trim off the 5’ end of the reverse read (reverse primer) -A: sequence to trim off the 3’ end of the reverse read (reverse complemented forward primer)

We’ll also add -m 50 to get rid of super short junky reads, --max-n 1 to get rid of reads that have any N’s in them and -n 2 so that cutadapt will remove multiple primer hits if there happens to be read-through. The --discard-untrimmed is also added so only reads that contain a primer will be kept ensuring we keep only valid amplicons. And finally a -q 15 will trim off low quality bases from the 3’ end.
# Run Cutadapt


```{r}
cutadapt_args <- c("-g", FWD, "-a", REV.RC, 
                   "-G", REV, "-A", FWD.RC,
                   "-n", 2, "--discard-untrimmed")


R1.flags <- paste("-g", FWD, "-a", REV.RC) 
# Trim REV and the reverse-complement of FWD off of R2 (reverse reads)
R2.flags <- paste("-G", REV, "-A", FWD.RC)

```

```{r}
for(i in seq_along(fnFs)) {
  system2(cutadapt, args = c(R1.flags, R2.flags, "-n", 2, # -n 2 required to remove FWD and REV from reads
                               "-o", fnFs.cut[i], "-p", fnRs.cut[i], # output files
                               fnFs.filtN[i], fnRs.filtN[i]))
}


```




# sanity check
# This should show no occurences in any of the orientations now
```{r}
rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs.cut[[1]]), 
      FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = fnRs.cut[[1]]), 
      REV.ForwardReads = sapply(REV.orients, primerHits, fn = fnFs.cut[[1]]), 
      REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnRs.cut[[1]]))

```



# remove any empty reads
```{r}
fnFs.clean <- fnFs.cut %>% str_replace("_R1","_clean_R1")
fnRs.clean <- fnRs.cut %>% str_replace("_R2","_clean_R2")
filterAndTrim(fnFs.cut,fnFs.clean,fnRs.cut,fnRs.clean,minLen = 100,multithread=FALSE)
```


############################
My own code 
######################
#####################

```{r}


fwd_files <- sort(list.files(path, pattern = "_R1_001.fastq", full.names = TRUE)) 
rev_files <- sort(list.files(path, pattern = "_R2_001.fastq", full.names = TRUE))


# It's also handy to have a vector of sample names, which in this case is everything up 
# until the first underscore, which is what our regular expression caputres.  You may
# also create this manually if you don't have too many samples
samples = str_extract(basename(fwd_files), "^[^_]+")


names(fwd_files) <- samples
names(rev_files) <- samples


fwd_primer <- "ACGCGHNRAACCTTACC"
rev_primer <- "ACGGGCRGTGWGTRCAA"
fwd_primer_rev <- as.character(reverseComplement(DNAStringSet(fwd_primer)))
rev_primer_rev <- as.character(reverseComplement(DNAStringSet(rev_primer)))
```


```{r}
count_primers <- function(primer, filename) {
  num_hits <- vcountPattern(primer, sread(readFastq(filename)), fixed = FALSE)
  return(sum(num_hits > 0))
}

count_primers(fwd_primer, fwd_files[[1]]) # 139018
count_primers(rev_primer, rev_files[[1]]) #135025
```





```{r}
# Create an output directory to store the clipped files
cut_dir <- file.path(path, "cutadapt")
if (!dir.exists(cut_dir)) dir.create(cut_dir)

fwd_cut <- file.path(cut_dir, basename(fwd_files))
rev_cut <- file.path(cut_dir, basename(rev_files))

names(fwd_cut) <- samples
names(rev_cut) <- samples

# It's good practice to keep some log files so let's create some
# file names that we can use for those 
cut_logs <- path.expand(file.path(cut_dir, paste0(samples, ".log")))

cutadapt_args <- c("-g", fwd_primer, "-a", rev_primer_rev, 
                   "-G", rev_primer, "-A", fwd_primer_rev,
                   "-n", 2, "--discard-untrimmed")

# Loop over the list of files, running cutadapt on each file.  If you don't have a vector of sample names or 
# don't want to keep the log files you can set stdout = "" to output to the console or stdout = NULL to discard
for (i in seq_along(fwd_files)) {
  system2(cutadapt, 
          args = c(cutadapt_args,
                   "-o", fwd_cut[i], "-p", rev_cut[i], 
                   fwd_files[i], rev_files[i]),
          stdout = cut_logs[i])  
}

# quick check that we got something
head(list.files(cut_dir))
```

